### PR TITLE
Support query auto timeout cancel on thriftserver

### PR DESCRIPTION
**Pull request by ulysses-you**
ulysses-you [Closes #451](https://github.com/yaooqinn/kyuubi/pull/451)    Target Issue Test Plan Feature [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=yaooqinn&utm_campaign=badge_info)

### _Why are the changes needed?_

Manual cherry-pick some Spark patch into Kyuubi.
1. [Support query auto timeout cancel on thriftserver](https://github.com/apache/spark/pull/29933)
2. [Add config to control if cancel invoke interrupt task on thriftserver](https://github.com/apache/spark/pull/30481)

In order to keep backward with early Spark version, we hard code the config key instead of refer to Spark SQLConf.

Note that, the exists timeout of operator (`kyuubi.operation.idle.timeout`) is to cancel that client has no access with engine. That said if a query run a long time and the client is alive, the query would not be cancelled. Then the new added config `spark.sql.thriftServer.queryTimeout` can handle this case.

### _How was this patch tested?_
Add new test.

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @pan3793 @yaooqinn

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/Constants.scala [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.engine.spark
+
+object Constants {
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/Constants.scala:20]: We should define `kyuubi.server.query.timeout` at server-side rather engine-side.

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/operation/SparkOperationSuite.scala [hunk 2]
```diff
@@ -746,4 +751,96 @@ class SparkOperationSuite extends WithSparkSQLEngine with JDBCTests {
       assert(tFetchResultsResp.getStatus.getStatusCode === TStatusCode.SUCCESS_STATUS)
     }
   }
+
+  test("SPARK-26533: Support query auto timeout cancel on thriftserver - setQueryTimeout") {
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/operation/SparkOperationSuite.scala:755]: we need to move these test to the superclass to test both the server and engine.

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala [hunk 3]
```diff
@@ -32,9 +34,25 @@ class ExecuteStatement(
...
     session: Session,
     protected override val statement: String,
-    override val shouldRunAsync: Boolean)
+    override val shouldRunAsync: Boolean,
+    queryTimeout: Long)
   extends SparkOperation(spark, OperationType.EXECUTE_STATEMENT, session) with Logging {
 
+  // If a timeout value `queryTimeout` is specified by users and it is smaller than
+  // a global timeout value, we use the user-specified value.
+  // This code follows the Hive timeout behaviour (See #29933 for details).
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala:43]: how about replace `#29933` by `SPARK-26533`?

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/Constants.scala (hunk 1 above)
Comment (pan3793) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/Constants.scala:20]: Users can't infer time unit from `kyuubi.server.query.timeout`, and there are some configurations defined in second and others defined in millisecond, should we make the configuration name more clear?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 4]
```diff
@@ -502,6 +502,27 @@ object KyuubiConf {
       .timeConf
       .createWithDefault(Duration.ofSeconds(5).toMillis)
 
+  val OPERATION_FORCE_CANCEL: ConfigEntry[Boolean] =
+    buildConf("operation.interruptOnCancel")
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:506]: nit: we don't do camel style for configs in Kyuubi just `interrupt.on.cancel` for consistiency

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 5]
```diff
@@ -502,6 +502,27 @@ object KyuubiConf {
...
+      .doc("Set a query duration timeout in seconds in Kyuubi. If the timeout is set to " +
+        "a positive value, a running query will be cancelled automatically when the timeout is " +
+        "exceeded, otherwise the query continues to run till completion. If timeout values are " +
+        "set for each statement via `java.sql.Statement.setQueryTimeout` and they are smaller " +
+        "than this configuration value, they take precedence. If you set this timeout and prefer " +
+        "to cancel the queries right away without waiting task to finish, consider enabling " +
+        s"${OPERATION_FORCE_CANCEL.key} together.")
+      .version("1.2.0")
+      .timeConf
+      .createWithDefault(Duration.ofSeconds(0).toMillis)
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:524]: is it just 0L?

File: kyuubi-main/src/main/scala/org/apache/kyuubi/operation/ExecuteStatement.scala [hunk 6]
```diff
@@ -71,12 +73,24 @@ class ExecuteStatement(
...
       val req = new TExecuteStatementReq(remoteSessionHandle, statement)
       req.setRunAsync(shouldRunAsync)
+      req.setQueryTimeout(getQueryTimeout)
       val resp = client.ExecuteStatement(req)
       verifyTStatus(resp.getStatus)
       _remoteOpHandle = resp.getOperationHandle
     } catch onError()
   }
 
+  private def getQueryTimeout: Long = {
```
Comment (yaooqinn) [inline kyuubi-main/src/main/scala/org/apache/kyuubi/operation/ExecuteStatement.scala:83]: move this to sessionManager?

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala [hunk 7]
```diff
@@ -32,9 +35,13 @@ class ExecuteStatement(
     spark: SparkSession,
     session: Session,
     protected override val statement: String,
-    override val shouldRunAsync: Boolean)
+    override val shouldRunAsync: Boolean,
+    queryTimeout: Long)
   extends SparkOperation(spark, OperationType.EXECUTE_STATEMENT, session) with Logging {
 
+  private val forceCancel =
+    session.sessionManager.getConf.get(KyuubiConf.OPERATION_FORCE_CANCEL)
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala:43]: I guess we should get it from  `spark.conf`

**Comment by yaooqinn** [comment]
a local test run is needed when new configs are comming

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala (hunk 7 above)
Comment (ulysses-you) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala:43]: It's a global config, can only passed by jdbc url.

File: kyuubi-main/src/main/scala/org/apache/kyuubi/operation/ExecuteStatement.scala (hunk 6 above)
Comment (ulysses-you) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/operation/ExecuteStatement.scala:83]: We only do the timeout check only in `ExecteStatement` operation, no stronge opinion to move that..

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/IndividualSparkSuite.scala [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+import org.apache.spark.scheduler.{SparkListener, SparkListenerTaskEnd}
+import org.apache.spark.sql.SparkSession
+import org.scalatest.concurrent.PatienceConfiguration.Timeout
+import org.scalatest.time.SpanSugar._
+
+import org.apache.kyuubi.KyuubiFunSuite
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.operation.JDBCTestUtils
+
+class SparkEngineSuites extends KyuubiFunSuite {
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/IndividualSparkSuite.scala:33]: This new test class is for easily test which need modify the global config.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 9]
```diff
@@ -502,6 +502,27 @@ object KyuubiConf {
...
+  val OPERATION_FORCE_CANCEL: ConfigEntry[Boolean] =
+    buildConf("operation.interrupt.on.cancel")
+      .doc("When true, all running tasks will be interrupted if one cancels a query. " +
+        "When false, all running tasks will remain until finished.")
+      .version("1.2.0")
+      .booleanConf
+      .createWithDefault(true)
+
+  val OPERATION_QUERY_TIMEOUT: ConfigEntry[Long] =
+    buildConf("operation.query.timeout")
```
Comment (ulysses-you) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:514]: Decide to hide the Spark thing that in `SparkThriftServer,` so define the config with Kyuubi style @pan3793

**Comment by ulysses-you** [comment]
After the refector, provide some summary for reviewer:

1. Add the query timeout at Kyuubi Server and that can be modified by session global config and statement query tiemout.
2. Engine side just receive the query timeout and to do the timeout check.
3. Add a interrupt flag to Spark local property with `true` default that is also a session global config.
4. In order to add engine test in one class eaily, add a new class `IndividualSparkSuite` and change some related test code place.

File: kyuubi-main/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala [hunk 10]
```diff
@@ -49,6 +50,17 @@ class KyuubiOperationManager private (name: String) extends OperationManager(nam
     tSessionHandle
   }
 
+  private def getQueryTimeout(statementTimeout: Long): Long = {
+    // If a timeout value `queryTimeout` is specified by users and it is smaller than
+    // a session timeout value, we use the user-specified value.
+    val sessionTimeout = getConf.get(KyuubiConf.OPERATION_QUERY_TIMEOUT)
+    if (sessionTimeout > 0 && (statementTimeout <= 0 || sessionTimeout < statementTimeout)) {
```
Comment (yaooqinn) [inline kyuubi-main/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala:57]: what if sessionTimeout=-1 and statementTimeout=-1?

Comment (ulysses-you) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala:57]: nothing happened, that only add timeout checkout when timeout > 0.

File: kyuubi-main/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala [hunk 11]
```diff
@@ -49,6 +50,17 @@ class KyuubiOperationManager private (name: String) extends OperationManager(nam
     tSessionHandle
   }
 
+  private def getQueryTimeout(statementTimeout: Long, session: Session): Long = {
+    // If a timeout value `queryTimeout` is specified by users and it is smaller than
+    // a session timeout value, we use the user-specified value.
+    val sessionTimeout = session.sessionConf.get(KyuubiConf.OPERATION_QUERY_TIMEOUT)
```
Comment (yaooqinn) [inline kyuubi-main/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala:56]: getConf.get(KyuubiConf.OPERATION_QUERY_TIMEOUT)?

File: kyuubi-main/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala [hunk 12]
```diff
@@ -49,6 +50,17 @@ class KyuubiOperationManager private (name: String) extends OperationManager(nam
     tSessionHandle
   }
 
+  private def getQueryTimeout(statementTimeout: Long, session: Session): Long = {
```
Comment (yaooqinn) [inline kyuubi-main/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala:53]: we don't need session as parameter here

File: kyuubi-main/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala [hunk 13]
```diff
@@ -49,6 +50,17 @@ class KyuubiOperationManager private (name: String) extends OperationManager(nam
...
+    // If a timeout value `queryTimeout` is specified by users and it is smaller than
+    // a session timeout value, we use the user-specified value.
+    val sessionTimeout = getConf.get(KyuubiConf.OPERATION_QUERY_TIMEOUT)
```
Comment (yaooqinn) [inline kyuubi-main/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala:56]: nit: rename session timeout to `systemQueryTimeout` and `statementTimeout` to `clientQueryTimeout`

File: kyuubi-main/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala (hunk 12 above)
Comment (ulysses-you) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala]: ah, good catch ~

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala [hunk 14]
```diff
@@ -100,4 +108,26 @@ class ExecuteStatement(
       executeStatement()
     }
   }
+
+  private def addTimeoutMonitor(): Unit = {
+    if (queryTimeout > 0) {
+      val timeoutExecutor = Executors.newSingleThreadScheduledExecutor()
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala:114]: Use `NamedThreadFactory` to create thread

File: kyuubi-main/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala [hunk 15]
```diff
@@ -49,6 +50,18 @@ class KyuubiOperationManager private (name: String) extends OperationManager(nam
     tSessionHandle
   }
 
+  private def getQueryTimeout(clientQueryTimeout: Long): Long = {
+    // If a timeout value `queryTimeout` is specified by users and it is smaller than
+    // a session timeout value, we use the user-specified value.
```
Comment (yaooqinn) [inline kyuubi-main/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala:55]: nit: also update comments

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala (hunk 14 above)
Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala:114]: org.apache.kyuubi.util.ThreadUtils$#newDaemonSingleThreadScheduledExecutor

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 9 above)
Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:514]: I think Kyuubi style timeConf defined in ms? The comments here is in seconds.

Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:514]: Actually, I think we don't need specific time unit in kyuubi timeConf style, user should not care about what we use internal, they just define it via ISO-8601 duration string

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 16]
```diff
@@ -502,6 +502,27 @@ object KyuubiConf {
...
+      .doc("Set a query duration timeout in seconds in Kyuubi. If the timeout is set to " +
+        "a positive value, a running query will be cancelled automatically when the timeout is " +
+        "exceeded, otherwise the query continues to run till completion. If timeout values are " +
+        "set for each statement via `java.sql.Statement.setQueryTimeout` and they are smaller " +
+        "than this configuration value, they take precedence. If you set this timeout and prefer " +
+        "to cancel the queries right away without waiting task to finish, consider enabling " +
+        s"${OPERATION_FORCE_CANCEL.key} together.")
+      .version("1.2.0")
+      .timeConf
+      .createWithDefault(0L)
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:524]: `Duration.ZERO.toMillis` has better semantic

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/Constants.scala (hunk 1 above)
Comment (ulysses-you) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/Constants.scala]: @pan3793 
I see the point, but the exists conf already named with `*.timeout`.. How about enhance the `kyuubi-defaults.conf.template` ? I guess it's useful if we write some common config into it.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 17]
```diff
@@ -502,6 +502,27 @@ object KyuubiConf {
...
+      .doc("When true, all running tasks will be interrupted if one cancels a query. " +
+        "When false, all running tasks will remain until finished.")
+      .version("1.2.0")
+      .booleanConf
+      .createWithDefault(true)
+
+  val OPERATION_QUERY_TIMEOUT: ConfigEntry[Long] =
+    buildConf("operation.query.timeout")
+      .doc("Set a query duration timeout in seconds in Kyuubi. If the timeout is set to " +
+        "a positive value, a running query will be cancelled automatically when the timeout is " +
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:516]: when the timeout is exceeded -> `when timeout` or `if timeout`

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 18]
```diff
@@ -502,6 +502,27 @@ object KyuubiConf {
...
+        "exceeded, otherwise the query continues to run till completion. If timeout values are " +
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:517]: , otherwise -> `. Otherwise, `

**Review by yaooqinn — APPROVED** [review]
LGTM, if the test flakiness fixed

**Comment by yaooqinn** [comment]
thanks, merged to master for v1.2.0 and branch 1.1 for v1.1.1

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 9 above)
Comment (yaooqinn) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:514]: Sorry, I missed this comment. What @pan3793 pointed out is true, we don't need to tell the unit of a time conf, it's Duration  ISO-8601 strings..

Here is a bug at caller side! the client timeout is conf, and the system timeout is ms... @ulysses-you

**Pull request closed without merging**